driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java

```diff
@@ -0,0 +1,212 @@
+ * - if the current timer times out, in which case the events are delivered asynchronously;
+ * - if the queue is full, in which case the events are delivered synchronously.
+ */
+class AbstractEventDebouncer<T> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractEventDebouncer.class);
+
+    private final AtomicReference<DeliveryAttempt> pendingDelivery = new AtomicReference<DeliveryAttempt>(null);
+
+    private final ReentrantLock eventsLock = new ReentrantLock();
+
+    private final ReentrantLock startLock = new ReentrantLock();
+
+    private final Condition hasStarted = startLock.newCondition();
+
+    private final ScheduledExecutorService executor;
+
+    private final DeliveryCallback<T> callback;
+
+    private final long delayMs;
+
+    private final BlockingQueue<T> events;
+
+    private volatile boolean started = false;
+
+    AbstractEventDebouncer(ScheduledExecutorService executor, DeliveryCallback<T> callback, long delayMs, int maxPendingEvents) {
+        this.executor = executor;
+        this.callback = callback;
+        this.delayMs = delayMs;
+        this.events = new ArrayBlockingQueue<T>(maxPendingEvents);
+    }
+
+    void start() {
+        logger.trace("Starting debouncer...");
+        startLock.lock();
+        try {
+            started = true;
+            hasStarted.signalAll();
+        } finally {
+            startLock.unlock();
+        }
+        logger.trace("Debouncer started");
+        eventsLock.lock();
+        try {
+            if (!events.isEmpty()) {
+                logger.trace("Events were accumulated before the debouncer started: scheduling delivery now");
+                scheduleNewDelivery();
+            }
+        } finally {
+            eventsLock.unlock();
+        }
+    }
+
+    void stop() {
+        logger.trace("Stopping debouncer...");
+        started = false;
+        while (true) {
+            DeliveryAttempt previous = cancelPendingDelivery();
+            if (pendingDelivery.compareAndSet(previous, null)) {
+                break;
+            }
+        }
+        logger.trace("Debouncer stopped");
+    }
+
+    void eventReceived(T event) {
+        checkNotNull(event);
+        logger.trace("Event received {}", event);
+        enqueueEvent(event);
+        scheduleNewDelivery();
+    }
+
+    void deliverEvents() throws InterruptedException {
+        final List<T> events = new ArrayList<T>();
+        int drained;
+        eventsLock.lock();
+        try {
+            drained = this.events.drainTo(events);
+        } finally {
+            eventsLock.unlock();
+        }
+        if (drained > 0) {
+            logger.trace("Delivering {} events", drained);
+            callback.deliver(events);
+        } else {
+            logger.trace("No events to deliver");
+        }
+    }
+
+    private void enqueueEvent(T event) {
+        eventsLock.lock();
+        try {
+            while (!events.offer(event)) {
```

Comment:
Events will accumulate until the queue is full, in which case the delivery is scheduled on the calling thread (back pressure).